common/transport/amqp/src/amqp.ts

```diff
@@ -145,44 +235,190 @@ export class Amqp {
+            let callback = this._connectionCallback;
+            this._connectionCallback = undefined;
+            this._fsm.transition('connecting_session', callback);
+          },
+          'connection_close': (context: EventContext) => {
+            debug('In the connection_close handler for the connecting state: ' + context.connection.name + ' with an indicated error of: ' + this._getErrorName(this._indicatedConnectionError));
+            let err = this._indicatedConnectionError;
+            let callback = this._connectionCallback;
+            this._indicatedConnectionError = undefined;
+            this._connectionCallback = undefined;
+            this._connectionCloseOccurred = true;
+            manageConnectionHandlers('removeListener');
+            this._fsm.transition('disconnected', callback, err);
+          },
+          'connection_error': (context: EventContext) => {
+            debug('In the connection_error handler for the connecting state: ' + context.connection.name + ' with error of: ' + this._getErrorName(context.connection.error));
+            debug('Deferring transition until the connection_close occurs.');
+            this._indicatedConnectionError = context.connection.error as AmqpError;
+          },
+          'error': (context: EventContext) => {
+            debug('In the \'error\' handler for the connecting state: ' + context.connection.name + ' with error of: ' + this._getErrorName(context.connection.error));
+            let callback = this._connectionCallback;
+            this._connectionCallback = undefined;
+            manageConnectionHandlers('removeListener');
+            this._fsm.transition('disconnected', callback, context.connection.error);
+          },
+          'disconnected': (context: EventContext) => {
+            debug('In the disconnect handler for the connecting state: ' + context.connection.name + ' with error of: ' + this._getErrorName(context.connection.error));
+            let callback = this._connectionCallback;
+            this._connectionCallback = undefined;
+            manageConnectionHandlers('removeListener');
+            this._fsm.transition('disconnected', callback, new errors.NotConnectedError('rhea: connection disconnected'));
+          },
+          '*': () => this._fsm.deferUntilTransition()
+        },
+        connecting_session: {
+          _onEnter: (connectCallback, result) => {
+            this._sessionCallback = connectCallback;
+            this._sessionResult = result;
+            //
+            // According to the rhea maintainers, one can depend on that fact that no actual network activity
+            // will occur until the nextTick() after the call to create_session.  Because of that, one can
+            // put the event handlers on the rhea session object returned from the create_session call and be assured
+            // that the listeners are in place BEFORE any possible events will be emitted on the session.
```

Comment:
This is neat, but why not add the handlers before calling create_session?  You seem to be going to great lengths to justify that the placement of this call shouldn't cause any problems, but I don't know why you put it after and not before.

Reply:
Because the create session is creating the emitter that the handlers are being placed on.

---
In reply to: [209383003](https://github.com/Azure/azure-iot-sdk-node/pull/344#discussion_r209383003) [](ancestors = 209383003)